### PR TITLE
Transaction monitoring error handling fixes, tests fixes

### PR DESCRIPTION
**Pull request by alonp99**
### Description
Elaborate on the subject, motivation, and context.

### Related issues
 * Provide a link to each related issue.

### Breaking changes
 * Describe the breaking changes that this pull request introduces.

### How these changes were tested
 * Describe the tests that you ran to verify your changes, including devices, operating systems, browsers and versions.

### Examples and references
 * Links, screenshots, and other resources related to this change.

### Checklist
- [] I have read the [contribution guidelines](CONTRIBUTING.md) of this project
- [] I have read the [style guidelines](STYLE_GUIDE.md) of this project
- [] I have performed a self-review of my own code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings and errors
- [] New and existing tests pass locally with my changes

___

## **Type**
enhancement, bug_fix

___

## **Description**
- Enhanced error handling across various components by standardizing the use of `Error` objects and improving error message clarity.
- Refactored `HttpExceptionFilter` to use external mappings and utility functions for better error code management.
- Updated logging configurations to include error stack traces for better debugging.
- Simplified error object constructions and removed redundant error checks in transaction controllers.
- Adjusted test configurations to run Jest tests sequentially to avoid potential race conditions.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>7 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>abstract-logger.ts</strong><dd><code>Update error method signature in abstract logger</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/common/abstract-logger/abstract-logger.ts
<li>Changed the <code>error</code> method signature to accept <code>Error</code> instead of <code>string</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-a6166fd3f62ab46e6f7cb60524798c38e4f43ea62ac56797a6e1230916786f3b">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>app-logger.service.ts</strong><dd><code>Enhance error handling in AppLoggerService</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/common/app-logger/app-logger.service.ts
<li>Modified <code>error</code> method to create an <code>Error</code> object if a string is passed.<br> <br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-55adf05f381da4a816b53b3a021eb41ea68ff389e9805e298cd4370a32449422">+5/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>HttpExceptions.filter.ts</strong><dd><code>Refactor HttpExceptionFilter with external error handling</code></dd></summary>
<hr>

services/workflows-service/src/common/filters/HttpExceptions.filter.ts
<li>Removed hardcoded error codes status mapping.<br> <li> Added external mapping and utility functions for error messages.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-b05e287a71f0aebc5b94e171f4d728c9bcc7451c329d7ed13df9738eb1fc5db0">+32/-30</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>winston-logger.ts</strong><dd><code>Enhance WinstonLogger with error stack support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/common/utils/winston-logger/winston-logger.ts
<li>Added error stack trace support in logger format.<br> <li> Adjusted <code>error</code> method to directly accept <code>Error</code> object.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-fd4a7890299f448e018f3d76daa70a5c81881da70ed7e1b8da421591f5e23e1d">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>errors.ts</strong><dd><code>Simplify ValidationError construction</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/errors.ts
<li>Simplified error object construction in <br><code>ValidationError.fromClassValidator</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-d229647efc5690d5796f358059115d6822d99b801b999f078feb05249659cd75">+1/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transaction.controller.external.ts</strong><dd><code>Streamline error handling in TransactionControllerExternal</code></dd></summary>
<hr>

services/workflows-service/src/transaction/transaction.controller.external.ts
- Removed redundant error handling in transaction creation methods.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-7f26bfa0ccf29a96c0ffd9edae350cfd67a33bd51637e539f06143ed159e9cb8">+8/-14</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transaction.service.ts</strong><dd><code>Enhance error handling in TransactionService</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/transaction/transaction.service.ts
<li>Enhanced error handling for transaction creation.<br> <li> Integrated new error message utility for Prisma errors.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-e3a24530c5cfef7de924d0eed08f4ab4a9864f3a7140ec551e8dbb885452f582">+15/-8</a>&nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Tests
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>transaction.controller.external.intg.test.ts</strong><dd><code>Update integration test for transaction error handling</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/transaction/transaction.controller.external.intg.test.ts
- Updated error message in test to reflect new error handling.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-133ea78a07500a44ba8129bfac104ac74b7dbed45e6d23e2e28403d344995a37">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Configure Jest to run tests sequentially</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/package.json
- Modified Jest test command to run tests sequentially.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2336/files#diff-65672d584fae78ff93600b1898b6b95bb50b5e51a25297cd36a840c37cd34b65">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

## Summary by CodeRabbit

- **New Features**
	- Updated error handling in logging services to accept both `Error` objects and strings.
	- Introduced a new function to handle error messages based on Prisma errors.
	- Expanded error message details in transaction conflict scenarios.
- **Bug Fixes**
	- Fixed conditional setting of `value` property in `ValidationError` class.
- **Refactor**
	- Improved logging format to include error stack tracing.
	- Updated error handling and response structures in transaction services to enhance clarity and detail.
- **Tests**
	- Modified a test case to provide more specific information on transaction existence errors.

### CODE REVIEW

File: services/workflows-service/package.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
     "start:watch": "nest start --watch",
     "start:debug": "nest start --debug --watch",
     "build": "nest build --path=tsconfig.build.json",
-    "test": "jest",
+    "test": "jest --runInBand",
```
Comment (alonp99) [inline services/workflows-service/package.json:19]: will run sequentially, concurrency fix here: https://github.com/ballerine-io/ballerine/pull/2337

File: services/workflows-service/src/common/abstract-logger/abstract-logger.ts [hunk 2]
```diff
@@ -7,6 +7,6 @@ export abstract class IAppLogger {
   abstract info(message: string, payload: LogPayload): void;
   abstract warn(message: string, payload: LogPayload): void;
   abstract debug(message: string, payload: LogPayload): void;
-  abstract error(message: string, payload: LogPayload): void;
+  abstract error(message: Error, payload: LogPayload): void;
```
Comment (Omri-Levy) [inline services/workflows-service/src/common/abstract-logger/abstract-logger.ts]: I'd expect the signature to be the same as `window.console.error` or at least what we use in the workflows service.
image

File: services/workflows-service/src/common/app-logger/app-logger.service.ts [hunk 3]
```diff
@@ -25,8 +25,11 @@ export class AppLoggerService implements LoggerService, OnModuleDestroy {
     this.logger.info(message, { ...this.getLogMetadata(), logData });
   }
 
-  error(message: string, logData: LogPayload = {}) {
-    this.logger.error(message, { ...this.getLogMetadata(), logData });
+  error(error: Error | string, logData: LogPayload = {}) {
```
Comment (Omri-Levy) [inline services/workflows-service/src/common/app-logger/app-logger.service.ts:28]: Should the type of `error` here be the same as in `abstract-logger.ts`?

File: services/workflows-service/src/transaction/transaction.service.ts [hunk 4]
```diff
@@ -43,19 +45,24 @@ export class TransactionService {
...
         if (
-          error instanceof Prisma.PrismaClientKnownRequestError &&
-          error.code === PRISMA_UNIQUE_CONSTRAINT_ERROR
+          (error as Prisma.PrismaClientKnownRequestError).name === 'PrismaClientKnownRequestError'
         ) {
-          errorToLog = new Error('Transaction already exists', { cause: error });
+          errorMessage = getErrorMessageFromPrismaError(
+            error as Prisma.PrismaClientKnownRequestError,
+          );
         } else {
```
Comment (Omri-Levy) [inline services/workflows-service/src/transaction/transaction.service.ts:59]: With this you won't need the assertions or depend on `name`.
```
export const isPrismaClientKnownRequestError = (error: unknown): error is Prisma.PrismaClientKnownRequestError => {
  return error instanceof Prisma.PrismaClientKnownRequestError;
}
```

**Review by Omri-Levy — APPROVED** [review]

File: services/workflows-service/src/transaction/transaction.service.ts (hunk 4 above)
Comment (alonp99) [reply services/workflows-service/src/transaction/transaction.service.ts:59]: `instanceof` was used before and didnt work, I guess it related to the Prisma way initiates the class

File: services/workflows-service/src/common/abstract-logger/abstract-logger.ts (hunk 2 above)
Comment (alonp99) [reply services/workflows-service/src/common/abstract-logger/abstract-logger.ts]: changed to error | string

**Pull request merged**